httpclient/src/main/java/feign/httpclient/ApacheHttpClient.java

```diff
@@ -153,13 +153,19 @@ HttpUriRequest toHttpUriRequest(Request request, Request.Options options) throws
   private ContentType getContentType(Request request) {
     ContentType contentType = ContentType.DEFAULT_TEXT;
     for (Map.Entry<String, Collection<String>> entry : request.headers().entrySet())
-    if (entry.getKey().equalsIgnoreCase("Content-Type")) {
-      Collection values = entry.getValue();
-      if (values != null && !values.isEmpty()) {
-        contentType = ContentType.create(entry.getValue().iterator().next(), request.charset());
-        break;
+      if (entry.getKey().equalsIgnoreCase("Content-Type")) {
+        Collection<String> values = entry.getValue();
+        if (values != null && !values.isEmpty()) {
+          String mimeType = values.iterator().next();
+          try {
+            contentType = ContentType.create(mimeType, request.charset());
+          } catch (IllegalArgumentException e) {
+            // MIME type may not contain reserved characters, so try to parse content type
+            contentType = ContentType.parse(mimeType);
```

Comment:
I meant that the comment "MIME type may not contain reserved characters, so
try to parse content type" has an if statement. this code shouldn't exist
unless it is tested, so I'd make a test that exercises that if statement

On Sun, Sep 4, 2016 at 1:43 AM, Andrey Stolyarchuk <notifications@github.com
> wrote:

> In httpclient/src/main/java/feign/httpclient/ApacheHttpClient.java
> <https://github.com/OpenFeign/feign/pull/453#discussion_r77440388>:
>
> > @@ -153,13 +153,19 @@ HttpUriRequest toHttpUriRequest(Request request, Request.Options options) throws
> >    private ContentType getContentType(Request request) {
> >      ContentType contentType = ContentType.DEFAULT_TEXT;
> >      for (Map.Entry<String, Collection<String>> entry : request.headers().entrySet())
> > -    if (entry.getKey().equalsIgnoreCase("Content-Type")) {
> > -      Collection values = entry.getValue();
> > -      if (values != null && !values.isEmpty()) {
> > -        contentType = ContentType.create(entry.getValue().iterator().next(), request.charset());
> > -        break;
> > +      if (entry.getKey().equalsIgnoreCase("Content-Type")) {
> > +        Collection<String> values = entry.getValue();
> > +        if (values != null && !values.isEmpty()) {
> > +          String mimeType = values.iterator().next();
> > +          try {
> > +            contentType = ContentType.create(mimeType, request.charset());
> > +          } catch (IllegalArgumentException e) {
> > +            // MIME type may not contain reserved characters, so try to parse content type
> > +            contentType = ContentType.parse(mimeType);
>
> Do you mean test this on correctness before trying to call create method?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/OpenFeign/feign/pull/453/files/e66d0bbe188548b8dbe4fa596ac13668cbdae5a0#r77440388>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAD61wOmXfy3tVZPqGESo_Sq7tVWWFRMks5qmbHWgaJpZM4Jz7yl>
> .
>